### DISCUSSION TITLE
configure max grouping size

### DISCUSSION THREAD

**Discussion by dandandy** [renovatebot/renovate#33003] [Suggest an Idea]
### Tell us more.

Hello Renovate team,

I'll like to raise a feature request to limit Renovate groups to a maximum size. 

For example:
```json
{
    "packageRules": [
        {
            "groupName": "Minor updates group",
            "matchUpdatesType": ["minor"],
            "maxGroupSize": 5
        }
    ]
}
```

So based on this example, when Renovate runs, if it finds there are <=5 minor updates it behaves as normal, but if it's >5 then Renovate will raise multiple pull requests. Eg. 10 minor upgrades will raise 2 PRs, 14 minor upgrades will raise 3 PR's, etc.

I would like this feature because I find that grouping is too all-inclusive (causing problems like one package causing the whole Renovate PR to go red and a large number of package upgrades in one PR is too high risk), but individual raises too many PRs (causing problems like merging many Renovate PRs clogging CD pipeline, and merges can cause merge conflicts between other renovate PRs in my lock files). 

So, I think being able to set a maximum grouping size will allow Renovate users to find a compromise between the two extremes.

If you'll allow this, I'm interested in contributing.

Thank you!

**Comment by rarkins** [comment]
I don't like this idea too much. How would you handle branch and PR naming? How would you try to keep the contents consistent and cycling/churning?

**Reply by dandandy** [reply]
Hey rarkins,

Good points. I just want to understand, is it the idea/premise itself you don't like, or is it the technical challenges? 

As for the technical challenges, a way to avoid the ones you mentioned could be to use consistent hashing to create sub-groups and consistently identify which packages would be in which sub-group to prevent churning. This would require changing my proposed config slightly—we introduce a config called `subGroupCount` instead of `maxGroupSize`.

```json
{
    "packageRules": [
        {
            "groupName": "Minor updates group",
            "matchUpdatesType": ["minor"],
            "subGroupCount": 5
        }
    ]
}
```

This config will make Renovate split "Minor updates group" into 5 sub groups. It will use consistent hashing on the package name to assign the package to a particular sub group (sub-group 1, 2, 3, 4 or 5), and the subgroup number will be in the PR title. When a new package update matches this rule Renovate will use the consistent hashing algorithm to identify which sub group it is in. If it is in a sub-group that already has a PR then it will update the PR. If the sub-group does not have a PR yet then it will create a PR for that sub group. 

This is different from my original suggestion because we aren't limiting the number of packages included in each PR based on the maxSize, but I think this still solves the same problem motivating this discussion by creating a way to split groupings into smaller sub-groups. 

What do you think of this?

**Reply by rarkins** [reply]
I don't really like the idea/premise. It seems like a hack/workaround. Although hashing in name could indeed produce consistent groupings, it would also be confusing/non-transparent to users. e.g. they'd ask "Why did X go in this group?" and the only answer would be "it's just because that's what the hash decided"
